Use an interface to define the structure of the LdapUser object

dmb-dz opened the pull request:
By using an interface it is now possible to write custom User implementations (specifiable via a config parameter) that e.g. only rely on authentication done via LDAP. This allows an overridden LdapUserprovider to make use of a custom user manager (e.g. FOSUserBundle).

This PR is completely BC with older versions of this bundle by using your LdapUser implementation as a config default. If you find this PR useful, I could write a cookbook article explaining the specific usage. ;-)

dmb-dz committed:
Extract an interface for the ldapuser

dmb-dz committed:
Update provider to use the interface instead of the implementation

dmb-dz committed:
Add configuration option to provide the user model

in config.yml. Defaults to the internal user model.

dmb-dz committed:
Instantiate a user from the provided model in LdapUserProvider

dmb-dz committed:
Fix the inheritance chain of the LdapUserInterface

so that implementations only need to implement this interface.

johnkary commented on User/LdapUser.php:
```diff
@@ -101,9 +100,9 @@ public function eraseCredentials()
     }
 
-    public function isEqualTo(UserInterface $user)
+    public function isEqualTo(\Symfony\Component\Security\Core\User\UserInterface $user)
```
LdapUserInterface extends Symfony\Component\Security\Core\User\EquatableInterface, so shouldn't this signature match its implementation of isEqualTo()?   `public function isEqualTo(UserInterface $user)` ?

dmb-dz replied:
The signature matches the UserInterface expectation of the EquatableInterface. It doesn't matter if you import it (e.g. via use `\Symfony\Component\Security\Core\User\UserInterface`) or specify it directly in the function. I chose the later because it's the only place in the object where it's used.

BorisMorel replied:
I agree with the @dmb-dz solution.

johnkary commented:
I don't see any issues here and am +1 on this addition.

dmb-dz committed:
Add documentation to explain the use of a custom user manager

like added in PR #69 [1].

[1] https://github.com/BorisMorel/LdapBundle/pull/69

dmb-dz committed:
Update configuration example

with the changes provided by PR #69 [1].

[1] https://github.com/BorisMorel/LdapBundle/pull/69

dmb-dz commented:
> I could write a cookbook article explaining the specific usage.

Added the promised documentation in the latest commits.

dmb-dz committed:
Update documentation.

The encoder for the custom user entity needs to be set in security.yml.

dmb-dz commented:
> How do you override <service id="imag_ldap.security.user.provider" ?

By redefining it in `services.xml` of my bundle, e.g.:

``` xml
<?xml version="1.0" ?>
<!-- Acme/DemoBundle/Resources/config/services.xml -->
<container xmlns="http://symfony.com/schema/dic/services"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
    <services>
        <service id="imag_ldap.security.user.provider" class="Acme\DemoBundle\Security\User\Provider\LdapUserProvider">
          <argument type="service" id="imag_ldap.ldap_manager" />
          <argument type="service" id="fos_user.user_manager" />
          <argument type="service" id="validator" />
          <argument>%imag_ldap.model.user_class%</argument>
        </service>
    </services>
</container>
```

BorisMorel commented:
It's override the default service defined by LdapBundle ?
Great !

Thanks for your PR.

dmb-dz commented:
> It's override the default service defined by LdapBundle ?

Yep. One could also define another provider in `security.yml` and use that one later on the relevant section of the firewall if that's a problem. I could cover that example as well in the documentation.
